File: cpp/src/opensky_client.cpp
#include "opensky_client.hpp"

#include <cstdlib>
#include <iostream>
#include <sstream>
#include <vector>

#ifndef WIN32_LEAN_AND_MEAN
#define WIN32_LEAN_AND_MEAN
#endif
#ifndef NOMINMAX
#define NOMINMAX
#endif
#include <windows.h>
#include <winhttp.h>

namespace olb {
namespace {

std::wstring Utf8ToWide(const std::string& s) {
  if (s.empty()) {
    return {};
  }
  const int n = MultiByteToWideChar(CP_UTF8, 0, s.c_str(),
                                    static_cast<int>(s.size()), nullptr, 0);
  if (n <= 0) {
    return {};
  }
  std::wstring w(static_cast<size_t>(n), L'\0');
  MultiByteToWideChar(CP_UTF8, 0, s.c_str(), static_cast<int>(s.size()),
                      w.data(), n);
  return w;
}

bool ReadHttpResponseBody(HINTERNET hRequest, std::string* out) {
  out->clear();
  for (;;) {
    DWORD avail = 0;
    if (!WinHttpQueryDataAvailable(hRequest, &avail)) {
      return false;
    }
    if (avail == 0) {
      break;
    }
    std::vector<char> buf(static_cast<size_t>(avail));
    DWORD read = 0;
    if (!WinHttpReadData(hRequest, buf.data(),
                         static_cast<DWORD>(buf.size()), &read)) {
      return false;
    }
    out->append(buf.data(), read);
  }
  return true;
}

void SkipWs(const std::string& s, std::size_t* i) {
  while (*i < s.size() &&
         (s[*i] == ' ' || s[*i] == '\t' || s[*i] == '\r' || s[*i] == '\n')) {
    ++(*i);
  }
}

bool ParseJsonString(const std::string& s, std::size_t* i, std::string* out) {
  SkipWs(s, i);
  if (*i >= s.size() || s[*i] != '"') {
    return false;
  }
  ++(*i);
  out->clear();
  while (*i < s.size()) {
    const char c = s[*i];
    if (c == '"') {
      ++(*i);
      return true;
    }
    if (c == '\\' && *i + 1 < s.size()) {
      out->push_back(s[*i + 1]);
      *i += 2;
      continue;
    }
    out->push_back(c);
    ++(*i);
  }
  return false;
}

bool ParseJsonNull(const std::string& s, std::size_t* i) {
  SkipWs(s, i);
  if (s.compare(*i, 4, "null") == 0) {
    *i += 4;
    return true;
  }
  return false;
}

bool ParseJsonBool(const std::string& s, std::size_t* i, bool* out) {
  SkipWs(s, i);
  if (s.compare(*i, 4, "true") == 0) {
    *i += 4;
    *out = true;
    return true;
  }
  if (s.compare(*i, 5, "false") == 0) {
    *i += 5;
    *out = false;
    return true;
  }
  return false;
}

bool ParseJsonNumber(const std::string& s, std::size_t* i, double* out,
                     bool* is_null) {
  SkipWs(s, i);
  *is_null = false;
  if (ParseJsonNull(s, i)) {
    *is_null = true;
    *out = 0.0;
    return true;
  }
  char* end = nullptr;
  *out = std::strtod(s.c_str() + *i, &end);
  if (end == s.c_str() + *i) {
    return false;
  }
  *i = static_cast<std::size_t>(end - s.c_str());
  return true;
}

bool ParseStateRow(const std::string& s, std::size_t* i, AircraftState* ac) {
  SkipWs(s, i);
  if (*i >= s.size() || s[*i] != '[') {
    return false;
  }
  ++(*i);

  // 0 icao24 (string)
  std::string icao;
  if (!ParseJsonString(s, i, &icao)) {
    return false;
  }
  SkipWs(s, i);
  if (*i >= s.size() || s[*i] != ',') {
    return false;
  }
  ++(*i);

  // 1 callsign (string|null)
  std::string callsign;
  SkipWs(s, i);
  if (ParseJsonNull(s, i)) {
    callsign.clear();
  } else if (!ParseJsonString(s, i, &callsign)) {
    return false;
  }
  SkipWs(s, i);
  if (*i >= s.size() || s[*i] != ',') {
    return false;
  }
  ++(*i);

  // 2 origin_country
  std::string country;
  if (!ParseJsonString(s, i, &country)) {
    return false;
  }

  // skip 3 time_position, 4 last_contact
  for (int k = 0; k < 2; ++k) {
    SkipWs(s, i);
    if (*i >= s.size() || s[*i] != ',') {
      return false;
    }
    ++(*i);
    double tmp = 0;
    bool is_null = false;
    if (!ParseJsonNumber(s, i, &tmp, &is_null)) {
      return false;
    }
  }

  // 5 lon, 6 lat, 7 baro_alt, 8 on_ground, 9 velocity, 10 true_track
  double lon = 0, lat = 0, alt = 0, vel = 0, track = 0;
  bool lon_null = false, lat_null = false, alt_null = false, vel_null = false,
       track_null = false;
  bool on_ground = false;

  auto next_num = [&](double* v, bool* nnull) -> bool {
    SkipWs(s, i);
    if (*i >= s.size() || s[*i] != ',') {
      return false;
    }
    ++(*i);
    return ParseJsonNumber(s, i, v, nnull);
  };
  if (!next_num(&lon, &lon_null) || !next_num(&lat, &lat_null) ||
      !next_num(&alt, &alt_null)) {
    return false;
  }
  SkipWs(s, i);
  if (*i >= s.size() || s[*i] != ',') {
    return false;
  }
  ++(*i);
  if (!ParseJsonBool(s, i, &on_ground)) {
    // sometimes null
    if (!ParseJsonNull(s, i)) {
      return false;
    }
    on_ground = false;
  }
  if (!next_num(&vel, &vel_null) || !next_num(&track, &track_null)) {
    return false;
  }

  // consume until end of this row array
  int depth = 1;
  while (*i < s.size() && depth > 0) {
    const char c = s[*i];
    if (c == '"') {
      std::string ignore;
      if (!ParseJsonString(s, i, &ignore)) {
        return false;
      }
      continue;
    }
    if (c == '[') {
      ++depth;
    } else if (c == ']') {
      --depth;
    }
    ++(*i);
  }

  if (icao.empty() || lon_null || lat_null) {
    return false;
  }
  ac->icao24 = icao;
  // trim callsign
  while (!callsign.empty() && callsign.back() == ' ') {
    callsign.pop_back();
  }
  ac->callsign = callsign;
  ac->origin_country = country;
  ac->lon_deg = lon;
  ac->lat_deg = lat;
  if (!alt_null) {
    ac->alt_m = alt;
  }
  ac->on_ground = on_ground;
  if (!vel_null) {
    ac->speed_mps = vel;
  }
  if (!track_null) {
    ac->heading_deg = track;
  }
  ac->valid = true;
  ac->source = "opensky";
  return true;
}

std::vector<AircraftState> ParseStatesJson(const std::string& json) {
  std::vector<AircraftState> out;
  const std::size_t key = json.find("\"states\"");
  if (key == std::string::npos) {
    return out;
  }
  std::size_t i = json.find('[', key);
  if (i == std::string::npos) {
    return out;
  }
  ++i;  // inside states array
  for (;;) {
    SkipWs(json, &i);
    if (i >= json.size()) {
      break;
    }
    if (json[i] == ']') {
      break;
    }
    if (json[i] == ',') {
      ++i;
      continue;
    }
    if (json.compare(i, 4, "null") == 0) {
      break;
    }
    AircraftState ac;
    if (!ParseStateRow(json, &i, &ac)) {
      // skip to next top-level element best-effort
      break;
    }
    if (ac.valid && !ac.on_ground) {
      out.push_back(std::move(ac));
    }
  }
  return out;
}

}  // namespace

void OpenSkyClient::SetBBox(double lamin, double lomin, double lamax,
                            double lomax) {
  lamin_ = lamin;
  lomin_ = lomin;
  lamax_ = lamax;
  lomax_ = lomax;
}

void OpenSkyClient::LoadBBoxFromEnv() {
  if (const char* v = std::getenv("OPENSKY_LAMIN")) {
    lamin_ = std::atof(v);
  }
  if (const char* v = std::getenv("OPENSKY_LOMIN")) {
    lomin_ = std::atof(v);
  }
  if (const char* v = std::getenv("OPENSKY_LAMAX")) {
    lamax_ = std::atof(v);
  }
  if (const char* v = std::getenv("OPENSKY_LOMAX")) {
    lomax_ = std::atof(v);
  }
}

std::vector<AircraftState> OpenSkyClient::FetchStates(std::string* error) {
  if (error) {
    error->clear();
  }

  std::ostringstream path;
  path << "/api/states/all?lamin=" << lamin_ << "&lomin=" << lomin_
       << "&lamax=" << lamax_ << "&lomax=" << lomax_;
  const std::wstring wPath = Utf8ToWide(path.str());

  HINTERNET hSession =
      WinHttpOpen(L"opensky-lattice-bridge/0.1", WINHTTP_ACCESS_TYPE_DEFAULT_PROXY,
                  WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS, 0);
  if (!hSession) {
    if (error) {
      *error = "WinHttpOpen failed";
    }
    return {};
  }
#ifndef WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_3
#define WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_3 0x00002000
#endif
  {
    DWORD protocols = WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_2 |
                      WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_3;
    WinHttpSetOption(hSession, WINHTTP_OPTION_SECURE_PROTOCOLS, &protocols,
                     sizeof(protocols));
  }

  HINTERNET hConnect =
      WinHttpConnect(hSession, L"opensky-network.org",
                     INTERNET_DEFAULT_HTTPS_PORT, 0);
  if (!hConnect) {
    if (error) {
      *error = "WinHttpConnect opensky-network.org failed";
    }
    WinHttpCloseHandle(hSession);
    return {};
  }

  HINTERNET hRequest = WinHttpOpenRequest(
      hConnect, L"GET", wPath.c_str(), nullptr, WINHTTP_NO_REFERER, nullptr,
      WINHTTP_FLAG_SECURE);
  if (!hRequest) {
    if (error) {
      *error = "WinHttpOpenRequest failed";
    }
    WinHttpCloseHandle(hConnect);
    WinHttpCloseHandle(hSession);
    return {};
  }

  if (!WinHttpSendRequest(hRequest, WINHTTP_NO_ADDITIONAL_HEADERS, 0,
                          WINHTTP_NO_REQUEST_DATA, 0, 0, 0) ||
      !WinHttpReceiveResponse(hRequest, nullptr)) {
    if (error) {
      *error = "OpenSky HTTP request failed";
    }
    WinHttpCloseHandle(hRequest);
    WinHttpCloseHandle(hConnect);
    WinHttpCloseHandle(hSession);
    return {};
  }

  DWORD status = 0;
  DWORD sz = sizeof(status);
  WinHttpQueryHeaders(hRequest,
                      WINHTTP_QUERY_STATUS_CODE | WINHTTP_QUERY_FLAG_NUMBER,
                      WINHTTP_HEADER_NAME_BY_INDEX, &status, &sz,
                      WINHTTP_NO_HEADER_INDEX);
  std::string body;
  ReadHttpResponseBody(hRequest, &body);
  WinHttpCloseHandle(hRequest);
  WinHttpCloseHandle(hConnect);
  WinHttpCloseHandle(hSession);

  if (status == 429) {
    if (error) {
      *error = "opensky rate limited (429) - slow poll or add optional account";
    }
    return {};
  }
  if (status != 200) {
    if (error) {
      *error = "opensky HTTP " + std::to_string(status);
    }
    return {};
  }

  auto states = ParseStatesJson(body);
  std::cerr << "[olb] opensky fetched airborne=" << states.size()
            << " bbox=[" << lamin_ << "," << lomin_ << " .. " << lamax_ << ","
            << lomax_ << "]\n";
  return states;
}

}  // namespace olb

File: bridge/opensky_client.py
"""OpenSky Network ADS-B state client (public REST)."""

from __future__ import annotations

import math
import os
import time
from dataclasses import dataclass
from typing import List, Optional, Tuple

import requests


@dataclass
class AircraftState:
    icao24: str
    callsign: str = ""
    origin_country: str = ""
    lat_deg: float = 0.0
    lon_deg: float = 0.0
    alt_m: Optional[float] = None
    heading_deg: float = 0.0
    speed_mps: float = 0.0
    on_ground: bool = False
    valid: bool = False
    source: str = "opensky"


class OpenSkyClient:
    """Poll OpenSky /api/states/all for a geographic bbox."""

    def __init__(
        self,
        lamin: Optional[float] = None,
        lomin: Optional[float] = None,
        lamax: Optional[float] = None,
        lomax: Optional[float] = None,
        user: Optional[str] = None,
        password: Optional[str] = None,
        timeout_s: float = 30.0,
    ) -> None:
        # Default: Dallas / DFW area
        self.lamin = float(os.environ.get("OPENSKY_LAMIN", lamin if lamin is not None else 32.5))
        self.lomin = float(os.environ.get("OPENSKY_LOMIN", lomin if lomin is not None else -97.3))
        self.lamax = float(os.environ.get("OPENSKY_LAMAX", lamax if lamax is not None else 33.1))
        self.lomax = float(os.environ.get("OPENSKY_LOMAX", lomax if lomax is not None else -96.4))
        self.user = (user or os.environ.get("OPENSKY_USER", "")).strip() or None
        self.password = (password or os.environ.get("OPENSKY_PASSWORD", "")).strip() or None
        self.timeout_s = timeout_s

    def fetch_states(self) -> Tuple[List[AircraftState], Optional[str]]:
        """Returns (states, error_message)."""
        url = "https://opensky-network.org/api/states/all"
        params = {
            "lamin": self.lamin,
            "lomin": self.lomin,
            "lamax": self.lamax,
            "lomax": self.lomax,
        }
        auth = (self.user, self.password) if self.user and self.password else None
        try:
            r = requests.get(url, params=params, auth=auth, timeout=self.timeout_s)
        except requests.RequestException as e:
            return [], f"opensky request failed: {e}"
        if r.status_code == 429:
            return [], "opensky rate limited (429) - slow poll or add OPENSKY_USER/PASSWORD"
        if r.status_code != 200:
            return [], f"opensky HTTP {r.status_code}: {r.text[:200]}"
        body = r.json()
        raw = body.get("states") or []
        out: List[AircraftState] = []
        for row in raw:
            ac = _parse_state_row(row)
            if ac and ac.valid and not ac.on_ground:
                out.append(ac)
        return out, None


def _parse_state_row(row: list) -> Optional[AircraftState]:
    # OpenSky state vector indices:
    # 0 icao24, 1 callsign, 2 origin_country, 5 lon, 6 lat, 7 baro_alt,
    # 8 on_ground, 9 velocity, 10 true_track, ...
    if not isinstance(row, list) or len(row) < 11:
        return None
    icao = (row[0] or "").strip()
    if not icao:
        return None
    lon = row[5]
    lat = row[6]
    if lon is None or lat is None:
        return None
    alt = row[7]
    vel = row[9]
    track = row[10]
    return AircraftState(
        icao24=icao,
        callsign=(row[1] or "").strip(),
        origin_country=(row[2] or "").strip(),
        lat_deg=float(lat),
        lon_deg=float(lon),
        alt_m=float(alt) if alt is not None else None,
        heading_deg=float(track) if track is not None else 0.0,
        speed_mps=float(vel) if vel is not None else 0.0,
        on_ground=bool(row[8]),
        valid=True,
        source="opensky",
    )


def sim_states(n: int = 3) -> List[AircraftState]:
    """Synthetic ADS-B tracks near Dallas for sandbox smoke."""
    t = time.time()
    lat0, lon0 = 32.8990, -97.0403  # DFW
    out: List[AircraftState] = []
    for i in range(n):
        ang = (t / 40.0 + i * (2.0 * math.pi / max(n, 1))) % (2.0 * math.pi)
        dlat = (0.08 * math.cos(ang + i)) 
        dlon = (0.10 * math.sin(ang + i))
        icao = f"sim{i:03d}"
        out.append(
            AircraftState(
                icao24=icao,
                callsign=f"SIM{i+1:03d}",
                origin_country="United States",
                lat_deg=lat0 + dlat,
                lon_deg=lon0 + dlon,
                alt_m=3000.0 + 250.0 * i,
                heading_deg=(math.degrees(ang) + 90.0) % 360.0,
                speed_mps=120.0,
                on_ground=False,
                valid=True,
                source="sim",
            )
        )
    return out
